src/Symfony/Component/AutoMapper/AutoMapper.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+namespace Symfony\Component\AutoMapper;
+
+use Psr\Container\ContainerInterface;
+use Symfony\Component\AutoMapper\Exception\RuntimeException;
+use Symfony\Component\AutoMapper\Attributes\MapIf;
+use Symfony\Component\AutoMapper\Attributes\MapTo;
+use Symfony\Component\AutoMapper\Attributes\MapWith;
+use Symfony\Component\PropertyAccess\PropertyAccessorInterface;
+
+/**
+ * Object to object mapper.
+ *
+ * @experimental
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+final class AutoMapper implements AutoMapperInterface
+{
+    public function __construct(private ?PropertyAccessorInterface $propertyAccessor = null, private ?ContainerInterface $serviceLocator = null)
+    {
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function map(object $object, object|string|null $to = null): mixed
+    {
+        $refl = new \ReflectionClass($object);
+
+        if (!$to) {
+            $to = $this->getAttribute($refl, MapTo::class, true)->to;
+        }
+
+        $arguments = [];
+        if (is_object($to)) {
+            $toRefl = new \ReflectionClass(get_class($to));
+            $mapped = $to;
+        } else {
+            $toRefl = new \ReflectionClass($to);
+            $mapped = $toRefl->newInstanceWithoutConstructor();
+        }
+
+        $constructor = $toRefl->getConstructor();
+        foreach ($constructor?->getParameters() ?? [] as $parameter) {
+            $arguments[$parameter->getName()] = $parameter->isDefaultValueAvailable() ? $parameter->getDefaultValue() : null;
+        }
+
+        foreach ($refl->getProperties() as $property) {
+            $propertyName = $property->getName();
+            $mapTo = $this->getAttribute($property, MapTo::class)?->to ?? $propertyName;
+            if (!$toRefl->hasProperty($mapTo)) {
+                continue;
+            }
+
+            $value = $this->propertyAccessor ? $this->propertyAccessor->getValue($object, $propertyName) : $object->{$propertyName};
+            $mapIf = $this->getCallable($this->getAttribute($property, MapIf::class)?->if);
+            if (is_callable($mapIf) && false === $this->call($mapIf, $value, $object)) {
+                continue;
+            }
+
+            $mapWith = $this->getCallable($this->getAttribute($property, MapWith::class)?->with);
+            if (is_callable($mapWith)) {
+                $value = $this->call($mapWith, $value, $object);
+            }
+
+            if (is_object($value) && $to = $this->getAttribute(new \ReflectionClass(get_class($value)), MapTo::class)?->to) {
```

Comment:
This is probably why the end arg to `->map()` allows null. If there was an `A.category` property that is a `CategoryDto` object... we would look on the `CategoryDto` class for a `MapTo` to know what to map it to.

That works... though I was thinking that we might find the matching property on the target class and read its type-hint to know what we're going to. So, if we're matching from `A` -> `B`, look at `B.category` to see that is a `CategoryFooBar` type, and them know that we need to map with *that* as the target.

Reply:
I'd like not to introduce type reading as it makes everything way more complicated. It works just fine by explicitly saying we want mapping no?

Reply:
We can at least start this way and see how it feels... since, you're right, that it will add some code.